[로또] 김시우 미션 제출합니다

seate opened the pull request:
## 추가 설명

### 추가 구현 기능 설명
1. **bean 등록을 위한 project scan 기능**
    - `@Controller`, `@Service`, `@Repository` 어노테이션이 적용된 클래스를 찾아서 반환한다.
2. **bean container**
    - 지정한 클래스를 bean으로 등록하고, 의존성 주입을 통해 객체를 생성한다.
    - 의존성 주입 시 순환 참조를 감지하여 예외를 발생시킨다.
    - bean container에서 bean을 가져올 때, 싱글톤 패턴을 적용하여 동일한 객체를 반환한다.
    - bean container에서 bean을 가져올 때, 해당 class와 interface를 통해 가져올 수 있다.
3. **DI 기능**
    - reflection으로 instance를 생성하면서 bean container에서 bean을 가져와서 의존성 주입을 진행한다.
4. **front와 back의 통신 기능**
    - front와 back의 통신을 위한 기능을 구현하기 위해 front의 스레드와 back의 스레드를 분리하였다.
    - front에서 입력을 받아 back으로 전달하고, back에서 결과를 front로 응답한다.
    - front와 back이 요청, 응답하는 과정에서 back에서 에러가 발생하면 front로 에러 발생 여부와 에러 메시지를 전달한다.

### 고려했던 점
1. **view의 역할**
    - view의 역할을 일반적인 frontEnd의 역할로 정의하고, 그에 따른 역할을 하도록 구현했습니다.
    - 다만, view에서 json을 전송하고 dto에 mapping하는 과정은 현재 자바로 통일된 코드에서는 불필요한 연산만 진행하는 것이라 판단하여, 생략하고 dto로 대체했습니다.
2. **파싱과 검증**
    - 파싱과 검증을 어느 단계에서 진행할지, 같이 진행해야 하는지에 대한 고민이 있었습니다.
        - 파싱은 frontEnd가 입력을 받아 json으로 가공하여 backEnd에게 전송하듯이 view에서 진행했습니다.
        - 검증은 객체의 정합성을 보장하고, 응집성을 높이기 위하여 객체 내부에 검증 로직을 작성하고, 객체 생성 시 검증을 진행하도록 구현했습니다.
3. **실행 진입점과 front와 back의 관리 주체**
    - front와 back이 분리된 application을 구현하기 위해 frontApplication, backApplication을 작성하여 각각의 관리 주체로 삼았습니다. 그리고 main에서 frontApplication과 backApplication을 실행시켜, 일반적인 웹 애플리케이션과 유사하게 동작하도록 구현할 수 있었습니다.
4. **front와 back의 통신**
    - front에서 입력을 받아 back으로 전달하고, back에서 결과를 front로 응답합니다.
    - front와 back이 요청, 응답하는 과정에서 back에서 에러가 발생하면 front로 에러 발생 여부와 에러 메시지를 전달합니다.
5. **스레드 분리에 따른 제약사항**
    - 스레드를 분리함에 따라, 테스트 코드의 random 값의 임의 지정이 적용되지 않아, back의 스레드는 현재 스레드에서 진행하는 것으로 하고, front의 스레드만 분리하였습니다.
6. **enum의 활용**
    - enum을 활용하라는 지시사항에 따라, Lotto에 사용되는 config와 lottoResult의 당첨 정보를 enum으로 관리할 수 있도록 하였습니다. enum으로 관리함에 따라 여러 클래스와 지점들에서 일관성 있게 관리할 수 있고, 수정이 용이하며, 값이 어떤 의미를 가지는 지 알 수 있다는 장점이 있습니다.
7. **front에서의 에러 처리 및 재시도 로직**
    - 일반적인 웹 애플리케이션과 유사하게 back에서 에러가 발생했다는 응답이 오면, front에서 에러 메시지를 출력하고, front에서 재시도하는 로직을 실행하였습니다.
8. **재시도 로직에서의 에러 처리**
    - 현재 테스트 코드에서 console.readline()에서 발생하는 NoSuchElementException을 통해 프로그램을 종료하는 것을 이용하고 있습니다. 스레드가 분리된 상태에서는 frontApplication이 NoSuchElementException을 통해 종료되면 backApplication도 종료되도록 구현해야 했기 때문에, finally문을 통해 정상 종료일 때도, 비정상 종료일 때도 backApplication이 종료되도록 구현하였습니다.

nyatda0116 committed:
docs(README): 기능 명세서 작성

- 예상되는 기능 명세서 작성

nyatda0116 committed:
feat(Controller, Service, Repository): bean 대상으로 등록할 어노테이션 구현

- Controller, Service, Repository 어노테이션 구현
- 어노테이션마다 다른 기능은 없으나 구분을 위해 세 가지를 구현

nyatda0116 committed:
feat(BeanConfig, BeanConfigTest): bean container 기능 구현

- 현재는 bean으로 등록할 객체에 하나의 생성자만 허용

nyatda0116 committed:
feat(ProjectScanner): 프로젝트의 파일을 scan하여  class를 return하는 기능 구현

- project scan 기능을 통해 bean으로 등록할 객체 scan 가능

nyatda0116 committed:
feat(ApplicationContext, BeanConfigTest): projectScan 기능과 bean Container 동작을 총괄하는 기능 구현

- project를 scan하고, scan된 객체들을 bean Container에 등록하는 기능 구현

nyatda0116 committed:
feat(Telecomunnication, TelecomunnicationTest, CustomCommunicationCode, CustomCommunicationData): front와 back의 통신을 기능 구현

- front와 back의 각각의 단방향 queue를 통해서 통신하는 기능 구현

nyatda0116 committed:
feat(BackApplication, FrontController): 백 애플리케이션 총괄 기능 구현

- applicationContext 초기화 후 실행 및 frontController 실행 기능 총괄

nyatda0116 committed:
feat(LottoConfig): 로또 설정 관리 enum 구현

- 일관성, 의미 및 수정 편의성을 위해 설정을 Enum으로 관리

nyatda0116 committed:
feat(LottoNumber, CustomIllegalArgumentException): 로또 번호 일관적 검증을 위한 기능 구현

- 로또 번호를 검증 시, 일관적 검증을 위한 Integer를 대신할 LottoNumber 일급 컬렉션 구현

nyatda0116 committed:
feat(Lotto, LottoTest): 로또 도메인 구현

- 생성 시 로또 번호 개수 및 중복 검증
- 랜덤한 숫자로 로또 생성하는 기능 구현

nyatda0116 committed:
feat(Lottos, LottosTest): 구입한 로또들을 나타내는 class 구현

- 가격 검증 후 개수 측정하여 로또들 생성 후 반환 기능 구현

nyatda0116 committed:
feat(FrontController, LottoController, LottoResponseDTO, PurchasedLottoResponseDTOs, LottoService, PurchasedLottoRepository, BeanConfigTest, PurchasedLottoRepositoryTest): 로또 구매 기능 구현

- 로또 구매 후, 이후 요청을 위해 저장

nyatda0116 committed:
feat(FrontController, LottoController, SetPrizeLottoRequestDTO, , LottoService, PrizeLottoRepository, BeanConfigTest, PrizeLottoRepositoryTest): 당첨 로또 지정 기능 구현

- 당첨 로또 지정 기능 구현

nyatda0116 committed:
feat(FrontController, LottoController, LottoService, LottoResult, LottoResultTest, LottoResults, LottoResultRequestDTO, LottoResultResponseDTO, LottoResultResponseDTOs, BeanConfigTest, LottoResultTest, LottoStatus, LottoStatusTest): 보너스 번호 지정 및 수익률 계산 기능 구현

- 보너스 번호 지정 및 수익률 계산 기능 구현

nyatda0116 committed:
feat(LottoRequestParser, LottoRequestParserTest): 로또 구입 금액 파싱 기능 구현

nyatda0116 committed:
feat(LottoRequestParser, LottoRequestParserTest): 당첨 번호 파싱 기능 구현

nyatda0116 committed:
feat(LottoRequestParser, LottoRequestParserTest): 보너스 번호 파싱 기능 구현

nyatda0116 committed:
feat(LottoRequestView): 사용자 입력 기능 구현

nyatda0116 committed:
feat(LottoResponseView): 출력 기능 구현

nyatda0116 committed:
feat(FrontApplication): 프론트 애플리케이션 총괄 및 재시도 기능 구현

- 프론트 공통 재시도 기능과 에러 출력을 구현하기 위하여 함수형 프로그래밍 기법 사용

nyatda0116 committed:
feat(Application): 애플리케이션 총괄 및 프론트, 백 분리를 위하여 스레드 분리

- 백 애플리케이션은 랜덤 임의 지정 테스트 코드를 위하여 현재 스레드에서 분리하지 않음

nyatda0116 committed:
docs(readme): 추가 구현 기능 설명과 고려사항 추가

whatsgoodg commented on src/main/java/lotto/front/util/LottoRequestParser.java:
```diff
@@ -11,4 +13,15 @@ public static Integer parsePrice(String price) {
         }
     }
+
+    public static List<Integer> parsePrizeNumber(String prizeNumber) {
+        try {
+            return Stream.of(prizeNumber.split(","))
+                    .map(String::trim)
+                    .map(Integer::parseInt)
+                    .toList();
```
Integer 변환이면 Integer.valueOf()를 사용하는게 좋아보입니다. parseInt는 int로 반환하고 이를 다시 Integer로 반환하여 불필요한 작업이 발생합니다. 또한 작은 수는 valueOf() 매서드에서 캐싱 기능을 사용합니다.

seate replied:
내부 코드를 보았을 때, valueOf는 parseInt를 호출합니다. 그리고 parseInt가 int를 반환합니다. 제 생각에 반대로 알고 계신 것 같습니다.
```java
public static Integer valueOf(String s) throws NumberFormatException {
    return Integer.valueOf(parseInt(s, 10));
}
```

```java
public static int parseInt(String s) throws NumberFormatException {
    return parseInt(s,10);
}
```

그리고 valueOf가 캐싱 기능을 갖고 있는 건 맞지만, parseInt가 반환된 후에 다시 찾으므로, parseInt보다 성능이 좋지 못할 것으로 보입니다.
```java
@IntrinsicCandidate
public static Integer valueOf(int i) {
    if (i >= IntegerCache.low && i <= IntegerCache.high)
        return IntegerCache.cache[i + (-IntegerCache.low)];
    return new Integer(i);
}
```

seate replied:
```java
public static Integer valueOf(String s) throws NumberFormatException {
        return Integer.valueOf(parseInt(s, 10));
    }
```
내부 코드에서 parseInt 다음에 valueOf를 시도하기 때문에 Integer 변환에 불필요한 작업이 발생한다는 것은 틀린 것 같습니다.
다만, valueOf의 캐싱 기능은 존재하기 때문에 valueOf로 리팩토링하도록 하겠습니다.

whatsgoodg commented on src/main/java/lotto/front/view/LottoRequestView.java:
```diff
@@ -0,0 +1,43 @@
+        return new PurchaseLottoRequestDTO(purchasePrice);
+    }
+
+    public static SetPrizeLottoRequestDTO requestPrizeNumbers(UUID uuid) {
+        System.out.println(PRIZE_NUMBERS_REQUEST_MESSAGE);
+        List<Integer> prizeNumbers = LottoRequestParser.parsePrizeNumber(Console.readLine());
+        System.out.println();
+
+        return new SetPrizeLottoRequestDTO(uuid, prizeNumbers);
+    }
```
이전에 말했듯이, prizeNumbers 보단 winning 또는 drawn을 사용하여 직관적으로 다가올 수 있게 네이밍하는 것이 좋아보입니다.

whatsgoodg commented on src/main/java/lotto/front/view/LottoResponseView.java:
```diff
@@ -0,0 +1,82 @@
+
+public class LottoResponseView {
+
+    private static final String LOTTO_COUNT_RESPONSE_MESSAGE = "개를 구매했습니다.";
+    private static final String PRIZE_RESULT_RESPONSE_MESSAGE = "당첨 통계";
+    private static final String PRIZE_RESULT_MESSAGE_DELIMITER = "---";
+    private static final String PRIZE_RESULT_RESPONSE_PREFIX_FORMAT = "%d개 일치";
+    private static final String PRIZE_RESULT_RESPONSE_SECOND_PRIZE_FORMAT = ", 보너스 볼 일치";
+    private static final String PRIZE_RESULT_RESPONSE_SUFFIX_FORMAT = " (%s원) - %d개";
+    private static final String PROFIT_RATE_RESPONSE_MESSAGE = "총 수익률은 %.1f%%입니다.";
```
모두 Enum으로 빼서 편하게 관리하는 것은 어떠신가요?

seate commented on src/main/java/lotto/back/lotto/repository/PrizeLottoRepository.java:
```diff
@@ -0,0 +1,22 @@
+import lotto.back.global.annotation.Repository;
+import lotto.back.lotto.domain.Lotto;
+
+@Repository
+public class PrizeLottoRepository {
+
+    private final Map<UUID, Lotto> lottoContainer = new ConcurrentHashMap<>();
```
로또라는 도메인 특성 상, 당첨 번호가 로또 번호 자체와 달라질 일이 없다고 판단하여, Lotto 객체가 가지고 있는 validation을 활용할 수 있게 같은 객체로 활용하였습니다.

말씀하신 대로 분리하면, 요구사항 변경에 더 유연하게 대처할 수 있게 되지만, 
1. 사실 상 같은 객체를 따로 관리해야 하는 문제 
2. 정합성 검증 로직이 이분화되어 객체 자체의 정합성을 매번 비교해야 하는 문제
3. 요구사항이 변경될 가능성이 거의 없는 도메인의 특성
4. 요구사항이 변경되더라도 이분되어 변경될 가능성이 매우 적은 도메인의 특성
같은 부작용이 말씀하신 장점보다 크다고 생각되어 같은 객체로 관리하게 되었습니다.

설령 위와 같은 상황이 발생하더라도, 분리하는 것은 크게 소요가 필요한 작업이 아니라고 판단했습니다.

seate commented on src/test/java/lotto/back/lotto/repository/PurchasedLottosRepositoryTest.java:
```diff
@@ -0,0 +1,26 @@
+
+    private final PurchasedLottosRepository purchasedLottosRepository = new PurchasedLottosRepository();
+
+    @Test
+    void 기본_동작_테스트() {
+        // given
+        UUID uuid = UUID.randomUUID();
+        Lottos lottos = Lottos.purchase(uuid, 1000);
+
+        // when
+        purchasedLottosRepository.save(lottos);
+        Lottos savedLottos = purchasedLottosRepository.findById(uuid);
+
+        // then
+        Assertions.assertThat(savedLottos).isEqualTo(lottos);
+    }
```
Lottos 객체 자체를 비교하기 때문에, uuid와 price에 대한 검증 모두 수행됩니다.

seate commented on src/main/java/lotto/back/lotto/domain/Lottos.java:
```diff
@@ -0,0 +1,58 @@
+
+public class Lottos {
+
+    private final UUID uuid;
```
uuid라는 필드명이 uuid를 대입해야 한다는 것을 나타내준다고 생각하기 때문에 유지하겠습니다.

seate commented on src/main/java/lotto/global/status/LottoStatus.java:
```diff
@@ -0,0 +1,49 @@
+                .orElse(NONE);
+    }
+
+    public static List<LottoStatus> getAllByAscendingPrize() {
+        return List.of(FIFTH, FOURTH, THIRD, SECOND, FIRST);
+    }
```
말씀하신 방법이 더 좋은 것 같아, 수정하였습니다. 감사합니다.

nyatda0116 committed:
refactor(LottoRequestParser): parseInt에서 valueOf로 변경

- valueOf 메서드의 캐싱 기능 활용을 위해 변경

nyatda0116 committed:
refactor(LottoStatus): 보상 목록 가져오기 하드 코딩 refactoring

nyatda0116 committed:
refactor(Lottos, LottoServiceImpl, LottosTest, PurchasedLottosRepositoryTest): 로또 구매 로직 refactoring

- 객체가 완결성 있게 생성되게 하기 위해 uuid 생성 로직을 객체 내부에 포함
- 구매 로직은 객체의 책임이 아니라고 판단하여, 서비스 로직이 담당하도록 변경

nyatda0116 committed:
fix(LottoServiceImpl, ApplicationTest): 수익률 계산 로직 오버플로우 fix

- int 범위를 넘어가는 경우를 대비하기 위해 당첨금 합산 로직에서 long을 사용하도록 변경

nyatda0116 committed:
fix(Lottos, LottoServiceImpl, LottosTest, PurchasedLottosRepositoryTest, ApplicationTest): price 지정 오류 해결

- price를 지정해야 하는 곳에 count를 지정하던 오류 해결
